Make the proc-macro FfiConverter impls canonical

bendk opened the pull request:
* Split out the proc-macro code to generate `FfiConverter` implementations and added standalone macros for that.
* Removed the `FfiConverter` implementations from the Askama template code.  Instead, the templates now invoke the macros.
* Removed the blanket `Arc<T>` impl from `uniffi_core` and made it so we invoke a macro for each interface.

The reason for this is:

* I want to add some more functionality to `FfiConverter`, but don't want to implement it twice and test it two different ways.
* Removing the blanket `Arc<T>` implementation allows us to customize the `FfiConverter` code for each interface.  I think this makes #1457 easier.
* Everything goes through one code path, which should give us some more confidence as we migrate to proc-macros.  This change caused me to fix some bugs and add some features to them.

bendk commented on fixtures/uitests/tests/ui/fieldless_errors_used_in_callbacks_cant_have_fields.stderr:
```diff
@@ -1,11 +1,13 @@
+  | | );
+  | |_^
+  |
+  = note: this error originates in the macro `::uniffi::ffi_converter_flat_error_with_read` (in Nightly builds, run with -Z macro-backtrace for more info)
```
These error messages aren't very good, but they don't seem any worse that the previous ones.

jplatte commented on uniffi_macros/src/lib.rs:
```diff
@@ -144,6 +144,90 @@ pub fn include_scaffolding(component_name: TokenStream) -> TokenStream {
+        let content;
+        syn::braced!(content in input);
+        let variants =
+            <syn::punctuated::Punctuated<syn::Variant, syn::Token![,]>>::parse_terminated(
```
The exact type should be inferrable from the return expression.

```suggestion
            syn::punctuated::Punctuated::parse_terminated(
```

jplatte commented on uniffi_macros/src/lib.rs:
```diff
@@ -144,6 +144,90 @@ pub fn include_scaffolding(component_name: TokenStream) -> TokenStream {
+    record::record_ffi_converter_impl(&input.fields, &input.ident).into()
+}
+
+struct FfiConverterEnumInput {
+    ident: syn::Ident,
+    variants: syn::punctuated::Punctuated<syn::Variant, syn::Token![,]>,
+}
```
I think it would make sense to move this into the `enum_` module.

jplatte commented on uniffi_macros/src/lib.rs:
```diff
@@ -144,6 +144,90 @@ pub fn include_scaffolding(component_name: TokenStream) -> TokenStream {
     }.into()
 }
 
+struct FfiConverterRecordInput {
```
Same here.

jplatte reviewed (commented):
Took a quick peek at the macro code, this is not a comprehensive review.

bendk commented on uniffi_macros/src/error.rs:
```diff
@@ -84,27 +87,49 @@ pub(crate) fn flat_error_ffi_converter_impl(
-        unsafe impl<T> ::uniffi::FfiConverter<T> for #ident {
-            ::uniffi::ffi_converter_rust_buffer_lift_and_lower!(T);
+        unsafe impl ::uniffi::FfiConverter<crate::UniFfiTag> for #ident {
+            ::uniffi::ffi_converter_rust_buffer_lift_and_lower!(crate::UniFfiTag);
```
This changes the code to only implement `FfiConverter<crate::UniFffiTag>`, meaning other crates don't get an automatic implementation.  I made the change because:

- I want to move the interface `FfiConverter` impls to the generate code, but that can't be a blanket impl since `Arc<T>` is not a local type
- The blanket impl doesn't work for remote types and this is currently a requirement for the UDL-based code
- I think we can still make sharing types between crates ergonomic.  What if the requirement was that users needed to write `uniffi_use!(external_crate::ExternalType)`?  That seems reasonable to me.  I hope to have another PR, based on top of this, that implements this.

jplatte replied:
I'm still against changing the default. How about having UDL codegen communicate to the proc-macro that it wants this sort of impl? Something like

```rust
#[derive(uniffi::Enum)]
#[uniffi(local_impl)]
enum Foo { ... }
```

bendk replied:
Yes, we can do that.

bendk commented on uniffi_macros/src/error.rs:
```diff
@@ -143,24 +197,39 @@ fn error_metadata(
+            })
+        } else {
+            Err(lookahead.error())
+        }
```
This is my best guess at how to handle more attributes here, please tell me if there's a better way to do this.

jplatte replied:
Exactly what I would have written 👍🏼

bendk commented:
Pushed an update based on the comments from @jplatte:
 - Switched the macros to use `DeriveInput`, which removes the need for parsing code
 - Added code to control when the macros generate `FfiConverter` implementations for all `<T>` vs a specific `UniFfiTag`.

jplatte commented on uniffi_macros/src/object.rs:
```diff
@@ -29,7 +30,24 @@ pub fn expand_object(input: DeriveInput, module_path: Vec<String>) -> TokenStrea
+    let (impl_spec, _) = tag_handler.into_impl_and_tag_path("Interface", ident);
+    quote! {
+        #[doc(hidden)]
+        #impl_spec { }
```
```suggestion
        #[automatically_derived]
        #impl_spec { }
```

jplatte commented on uniffi_macros/src/util.rs:
```diff
@@ -221,3 +222,72 @@ pub fn either_attribute_arg<T: ToTokens>(a: Option<T>, b: Option<T>) -> syn::Res
+    }
+}
+
+impl TryFrom<AttributeArgs> for FfiConverterTagHandler {
```
Why use syn's `Meta` types instead of parsing the attributes using `syn::parse::Parse` like elsewhere? I already added a bunch of helpers to make that easier.

jplatte replied:
I guess this is also the reason you didn't do `tag = crate::Tag` syntax, because `syn::Meta` doesn't support that?

bendk replied:
I tried, but couldn't figure out a good way to get from a `TokenStream` to an `[Attribute]` slicte to use with `parse_uniffi_attributes()`.  Maybe we could handle this as a follow-up.

jplatte replied:
Yeah, I think I can take care of that next week.

mhammond reviewed (approved):
This is a great improvement, thanks.

bendk commented on uniffi/tests/ui/proc_macro_arc.stderr:
```diff
@@ -1,31 +1,19 @@
+   = note: required for `Arc<Foo>` to implement `FfiConverter<UniFfiTag>`
+   = note: this error originates in the attribute macro `uniffi::export` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error[E0277]: the trait bound `child::Foo: Interface<UniFfiTag>` is not satisfied
```
I think I'm missing something because both error messages don't seem very clear to me.  I'm happy to work on improving these as a followup though.
